src/shell/shell.cpp

```diff
@@ -308,49 +262,57 @@ void DOS_Shell::ParseLine(char *line)
 		// LOG_MSG("SHELL: Redirecting output to %s",
 		//         pipe_file.length() ? pipe_tempfile : out_file.c_str());
-		close_stdout(normalstdout);
-		open_console_device(!normalstdin && !in_file.length());
-		if (!get_pipe_status(out_file.length() ? out_file.c_str() : nullptr,
-		                     pipe_file.length() ? pipe_file.c_str() : nullptr,
-		                     pipe_tempfile, append, failed_pipe) &&
-		    normalstdout) {
-			// Read only file, open con again
-			open_console_device();
+		output_redirection = get_output_redirection(
+		        out_file.length() ? out_file.c_str() : nullptr,
+		        pipe_file.length() ? pipe_file.c_str() : nullptr,
+		        pipe_tempfile,
+		        append,
+		        failed_pipe);
+		if (output_redirection == InvalidFileHandle) {
 			if (!pipe_file.length()) {
 				WriteOut(MSG_Get(dos.errorcode == DOSERR_ACCESS_DENIED
 				                         ? "SHELL_ACCESS_DENIED"
 				                         : "SHELL_FILE_CREATE_ERROR"),
 				         out_file.length() ? out_file.c_str()
 				                           : "(unnamed)");
-				close_stdout();
-				open_stdout_as("nul");
+				DOS_OpenFile("nul", OPEN_READWRITE, &output_redirection, true);
 			}
 		}
-		close_stdin(!normalstdin && !in_file.length());
 	}
+
+	// Replace stdin and stdout with redirection
+	if (input_redirection != InvalidFileHandle) {
+		psp->SetFileHandle(STDIN, input_redirection);
+	}
+	if (output_redirection != InvalidFileHandle) {
+		psp->SetFileHandle(STDOUT, output_redirection);
+	}
+
 	/* Run the actual command */
 	DoCommand(line);
+
 	/* Restore handles */
-	if (in_file.length()) {
-		close_stdin();
-		open_console_device(normalstdin);
+	if (input_redirection != InvalidFileHandle) {
+		psp->SetFileHandle(STDIN, old_stdin);
+		DOS_CloseFile(input_redirection, true);
 	}
-	if (out_file.length() || pipe_file.length()) {
-		close_stdout();
-		open_console_device(!normalstdin);
-		open_console_device(normalstdout);
-		close_stdin(!normalstdin);
+	if (output_redirection != InvalidFileHandle) {
+		psp->SetFileHandle(STDOUT, old_stdout);
+		DOS_CloseFile(output_redirection, true);
 	}
+
 	if (pipe_file.length()) {
+		uint16_t pipe_handle = InvalidFileHandle;
 		// Test if file can be opened for reading
-		if (!failed_pipe &&
-		    (dummy = open_stdin_as(pipe_tempfile)) != failed_open) {
-			DOS_CloseFile(dummy);
-			close_stdin(normalstdin);
-			open_stdin_as(pipe_tempfile); // Open new stdin
-			ParseLine((char *)pipe_file.c_str());
-			close_stdin();
-			open_console_device(normalstdin);
+		if (!failed_pipe && DOS_OpenFile(pipe_tempfile, OPEN_READ, &pipe_handle, true)) {
+			assert(pipe_handle != InvalidFileHandle);
+			psp->SetFileHandle(STDIN, pipe_handle);
+			assert(pipe_file.length() < CMD_MAXLINE);
+			char mutable_pipe_file[CMD_MAXLINE];
+			safe_strcpy(mutable_pipe_file, pipe_file.c_str());
+			ParseLine(mutable_pipe_file);
+			psp->SetFileHandle(STDIN, old_stdin);
+			DOS_CloseFile(pipe_handle, true);
```

Comment:
`true` 😉 and please insert some newlines while you're at it, this block of ASCII is absolutely unreadable.

Reply:
Fixed